### PR TITLE
[WIP] Make CommonsChunkPlugin care about indirect children, too

### PR DESCRIPTION
**Pull request by ArcEglos**
**What kind of change does this PR introduce?**

Feature

**Did you add tests for your changes?**

Not yet, wasn't able to find where tests for the CommonsChunkPlugin would have to go or where are some already.

**If relevant, link to documentation update:**

Will update Docs as soon as I'm confident this change will make it in and works as intended.

**Summary**

If you use the CommonsChunkPlugin with the children flag enabled it currently only cares about direct children. This change makes it so it cares about all children that are not required from somewhere else than the common chunk.
[https://github.com/webpack/webpack/issues/3981](Issue)

**Does this PR introduce a breaking change?**

If the config deepChildren isn't set the behavior stays the same as before, it only changes if the flag gets set.

**Other information**

Please be gentle, as this is my first try to get a PR merged.

### CODE REVIEW

File: lib/optimize/CommonsChunkPlugin.js [hunk 1]
```diff
@@ -46,8 +47,20 @@ The available options are:
 		const asyncOption = this.async;
 		const minSize = this.minSize;
 		const ident = this.ident;
+		const deepChildren = this.deepChildren;
 		compiler.plugin("this-compilation", (compilation) => {
 			compilation.plugin(["optimize-chunks", "optimize-extracted-chunks"], (chunks) => {
+				function pushUsedChunks(usedChunks, commonChunk, flags) {
+					(commonChunk.chunks || []).forEach((chunk) => {
+						if(flags.asyncOption || chunk.parents.every((parentChunk) => parentChunk === commonChunk || usedChunks.indexOf(parentChunk) >= 0)) {
```
Comment (sokra) [inline lib/optimize/CommonsChunkPlugin.js]: We can only use chunks if the commonChunk is the only parent. When recursively expanding all parents of the chunk (while ignoring circles) there must be only the commonChunk left. This condition must be ensured.

And we need a test for this. To create such a case you can create an additional entry point which also references the chunk which contains the common module (reference it via name, elsewise the optimizer will optimize it)

**Comment by ArcEglos** [comment]
Changed the little things as requested. Also fixed two bugs I found in the way I searched for the children.

The idea behind how I walk through the children of the common chunk is to check for each chunk if all parents are either the common chunk itself or another chunk I already know i use. As this checking begins with the children of the common chunk and I push the chunks I use before checking its children I can always be sure that I don't add any chunks I shouldn't.
At least that's what I think. I hope I didn't miss something really obvious.

I also added a testcase where you suggested it. I have to admit I'm not entirely sure my test really tests the right thing, tough. Had a bit of trouble wrapping my head around how to test my case from the source itself.

**Comment by sokra** [comment]
@timse Could you help resolving this conflict?

**Comment by timse** [comment]
will do! (tonight :))

**Comment by drphelps** [comment]
@timse, @sokra, what is the status of this PR? Getting this merged would make me happy.

**Comment by timse** [comment]
Sorry lost track of it. Will check tonight!

**Comment by timse** [comment]
hum im kind of overwhelmed given the changes to the current module and the changeset :p

**Comment by ArcEglos** [comment]
I had the same problem. Maybe it is simpler if I make the changes again based on the new module after you PR. Then we wouldn't need to try to resolve the conflicts.
I can look into the new version in a few hours if you want.

**Comment by timse** [comment]
i pushed a branch can you check please @ArcEglos
and close this one if its ok and reopen the pr with the new branch?
see here: https://github.com/webpack/webpack/pull/4416

**Comment by ArcEglos** [comment]
@timse great, thanks. I'll look into it later in the day

**Comment by ArcEglos** [comment]
Sorry for the confusion I'm creating but I didn't know how to properly use the branch you created, so I copied over the new Version from it and added the missing pieces. @timse

**Comment by timse** [comment]
no worries

**Comment by timse** [comment]
it looks however as you still have conflicts against master here

**Comment by ArcEglos** [comment]
Ah yeah, thanks for reminding. Resolved them.

**Comment by timse** [comment]
the tests fail and there is a console.log left, 
you could just take the diff of `https://github.com/webpack/webpack/pull/4416/files` and it should work.

**Comment by ArcEglos** [comment]
Ah sorry, I have to admit I only checked the example. My bad. I'll look into it.

Yep, the changes in #4416 didn't take into account chunks that have several parents but still only depend on the commonsChunk if you expand the whole dependency tree.
The approach I chose goes down from the common chunk and for each children recursively looks, if all parents of the children are either the common chunk or are in the list of chunks known to only have the common chunk as (grand, etc) parents. These children are added to said list before walking further down. When finished walking the dependency tree we take this list and know that all these chunks only have the commonChunk as parent.

**Comment by timse** [comment]
got it,

can we document that somehow? and could we split up the code a bit?
something like this:

```
// filter to only those chunks whose parents all are affected
filterChildenWithAllAffectedParents(affectedChunks, childChunks) {
  return childChunks.filter(() => {
    return chunk.parents.length === 1 ||
      chunk.parents.every((parentChunk) => affectedChunks.indexOf(parentChunk) > -1);
  });
} 

// remove duplicates of "existing"
filterAlreadyExistingChunks(existing, potential) {
  return potential.filter(p => existing.indexOf(p) === -1);
}

// get all child chunks that are affected - meaning they could be moved to the commons chunk
getAffectedUnnamedChunksFlat(affectedChunks, childChunks) {
  if (!childChunks) {
    return [];
  }

  if(this.async) {
    return childChunks;
  }

  return filterChildenWithAllAffectedParents(affectedChunks, childChunks)
}

// get all chunks that can be put into the commons chunk
getAffectedUnnamedChunks(affectedChunks, childChunks) {
  const affectedChildChunks = this.filterAlreadyExistingChunks(
    affectedChunks, 
    this.getAffectedUnnamedChunksFlat(affectedChunks, childChunks)
  );
  if(!this.deepChildren) {
    return affectedChildChunks;
  }
  return childChunks.reduce((allAffectedChunks, childChunk) => {
    return this.filterAlreadyExistingChunks(
      allAffectedChunks,
      this.getAffectedUnnamedChunksFlat(allAffectedChunks, childChunk.chunks)
    );
  }, affectedChunks);
}
```

that way we have clear functions and its a bit easier to understand what is happening

**Comment by timse** [comment]
The last part could probably be made nicer.
Also could this lead to endless loops? What if some chunks are each others parents/childrens

**Comment by ArcEglos** [comment]
I haven't had the time yet to refactor the code and comment it better. Hope to find time tonight, but won't make any promises. Just answering some of your questions in advance:

About the endless loops: These shouldn't be a problem, as we a) stop walking the dependencies at chunks that have parents we don't like and b) stop walking if we find a chunk has already been added to the chunks we are sure to only have parents we like.
So no chunk should be visited via the same connection more than once.

Maintaining these two conditions are also the main hurdle to seperate the whole thing into two steps because both the hierarchy while walking the dependencies and the timing of it are relevant to avoid endless loops and wrong assertions if a chunk is ok to add. But there is probably still a way to break it up.

**Comment by ArcEglos** [comment]
I hope this removes the problems with the tests (at least they run at my machine).
I also added more documentation what happens why.
When it comes to first flattening the tree before filtering it, I'm not really convinced this is a good idea. (At least to me) It makes the subtle ordering and hierarchical problems even less clear when reading the code. But maybe I'm the only one with this opinion @timse - maybe the annotated version also clarifies what's going on. It's up to you to judge ;)

**Comment by Tom910** [comment]
Hey, 1 test failed. This is a cool feature that I would like to see

```
 1 failing

  1) ConfigTestCases commons-chunk-plugin move-to-grandparent should compile:
     Uncaught SyntaxError: Unexpected token [
      at Object.exports.runInThisContext (vm.js:53:16)
      at _require (/home/travis/build/webpack/webpack/test/ConfigTestCases.test.js:95:18)
      at /home/travis/build/webpack/webpack/test/ConfigTestCases.test.js:124:9
      at /home/travis/build/webpack/webpack/lib/Compiler.js:9:17985
      at Compiler.emitRecords (/home/travis/build/webpack/webpack/lib/Compiler.js:9:24097)
      at /home/travis/build/webpack/webpack/lib/Compiler.js:9:17387
      at /home/travis/build/webpack/webpack/lib/Compiler.js:9:23782
      at next (/home/travis/build/webpack/webpack/node_modules/tapable/lib/Tapable.js:154:11)
      at Compiler.<anonymous> (/home/travis/build/webpack/webpack/lib/performance/SizeLimitsPlugin.js:9:4844)
      at Compiler.applyPluginsAsyncSeries1 (/home/travis/build/webpack/webpack/node_modules/tapable/lib/Tapable.js:158:13)
      at Compiler.afterEmit (/home/travis/build/webpack/webpack/lib/Compiler.js:9:23432)
      at Compiler.<anonymous> (/home/travis/build/webpack/webpack/lib/Compiler.js:9:23301)
      at /home/travis/build/webpack/webpack/node_modules/async/dist/async.js:359:16
      at iteratorCallback (/home/travis/build/webpack/webpack/node_modules/async/dist/async.js:935:13)
      at /home/travis/build/webpack/webpack/node_modules/async/dist/async.js:843:16
      at /home/travis/build/webpack/webpack/node_modules/graceful-fs/graceful-fs.js:43:10
      at FSReqWrap.oncomplete (fs.js:82:15)
```

**Comment by maroshluskasc** [comment]
@markrileybot did you end up working on this? If not, I'd like to help finishing it as well.

**Comment by daggerjames** [comment]
wait to be merged

**Comment by brad-sobie** [comment]
Need this

**Comment by abdavid** [comment]
Any progress with this PR? @markrileybot @maroshluskasc

**Comment by PTC-JoshuaMatthews** [comment]
I am amazed that commonChunksPlugin hasn't always worked this way, always assumed I was doing something wrong causing code duplication across  lazy loaded modules. Just to be clear, should I wait for this feature instead of attempting to optimize my bundles through other means? I was considering following the webpack docs and using dynamic imports for libraries that are being duplicated, but if this feature will figure the same splitting functionality with normal imports then I'd rather wait. Can anyone verify if this PR does that?

**Comment by grydstedt** [comment]
Would love to see this merged 👍

**Comment by brad-sobie** [comment]
Pretty please

**Comment by modosc** [comment]
just chiming in here but happy to help in any way to get this cleaned up / merged (or to pick it up into a new pr if need be).

**Comment by wbern** [comment]
Would be great to have this one put in! 👍 All the way.

**Comment by ljcrapo** [comment]
I have fixed all issues with this pull request and submitted another pull request:
https://github.com/webpack/webpack/pull/5764

**Comment by STRML** [comment]
This can be closed now that #5764 is merged & 3.7.0 is out! 🎉

**Comment by stevewillard** [comment]
@sokra BTW you can close this PR

**Pull request closed without merging**